base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -472,99 +451,108 @@ fn get_db_encryption(
+    // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
+    let pbkdf_params_version = 1u8;
+    let pbkdf_algorithm = argon2::Algorithm::Argon2id;
+    let pbkdf_version = argon2::Version::V0x13;
+    let pbkdf_params = argon2::Params::new(
+        46 * 1024,
+        1,
+        1,
+        Some(size_of::<Key>()), // for `XChaCha20-Poly1305`
+    )
+    .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
+    let main_key_domain = b"wallet_main_key_encryption".to_vec();
+
+    let main_key = match (secondary_key_version, secondary_key_salt, encrypted_main_key) {
+        // Encryption is not set up yet
+        (None, None, None) => {
+            // Generate a high-entropy main key
+            let mut main_key = WalletMainEncryptionKey::from(vec![0u8; size_of::<Key>()]);
+            let mut rng = OsRng;
+            rng.fill_bytes(main_key.reveal_mut());
+
+            // We'll be encrypting the main key shortly, so keep a clone around
+            let main_key_clone = main_key.clone();
+
+            // Derive the secondary key from the user's passphrase and a high-entropy salt
+            let secondary_key_salt = SaltString::generate(&mut rng);
+            let mut secondary_key = WalletSecondaryEncryptionKey::from(SafeArray::default());
+            argon2::Argon2::new(pbkdf_algorithm, pbkdf_version, pbkdf_params)
+                .hash_password_into(
                     passphrase.reveal(),
-                    Some(argon2::Algorithm::Argon2id.ident()),
-                    Some(argon2::Version::V0x13 as Decimal), // the API requires the numerical version representation
-                    params_passphrase,
-                    &passphrase_salt,
+                    secondary_key_salt.as_bytes(),
+                    secondary_key.reveal_mut(),
                 )
-                .map_err(|e| WalletStorageError::AeadError(e.to_string()))?
-                .to_string();
-
-            // Hash the passphrase to produce a ChaCha20-Poly1305 key
-            let encryption_salt = SaltString::generate(&mut OsRng);
-
-            // insert passphrase hash and encryption salt on the wallet db
-            WalletSettingSql::new(DbKey::PassphraseHash, passphrase_hash).set(conn)?;
-            WalletSettingSql::new(DbKey::EncryptionSalt, encryption_salt.to_string()).set(conn)?;
-
-            encryption_salt.to_string()
+                .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
+
+            // Use the secondary key to encrypt the main key, authenticating with the version to mitigate mismatch
+            // attacks
+            let main_key_cipher = XChaCha20Poly1305::new(Key::from_slice(secondary_key.reveal()));
+            let mut aad = main_key_domain;
+            aad.push(pbkdf_params_version);
+            let encrypted_main_key =
+                encrypt_bytes_integral_nonce(&main_key_cipher, aad, Hidden::hide(main_key.reveal().clone()))
```

Comment:
Encryption is made in-place on `main_key`, right ? If so, we would not need a `Hidden::hide` to hide its contents away.

Reply:
I considered writing a separate in-place function in the hopes that it would play nicely with `SafeArray`, but it proved to be more of a hassle.